pkg/controller/hosts_reconciler.go

```diff
@@ -0,0 +1,230 @@
+	for _, host := range hostsToProcess {
+		r.reconcileHost(host)
+	}
+	//r.updateDownedHosts(object, hostReport)
+}
+
+// HostAdded handles host addition event
+func (r *HostReconciler) HostAdded(host *southbound.Host) {
+	r.reconcileHost(host)
+}
+
+// HostDeleted handles host deletion event
+func (r *HostReconciler) HostDeleted(host *southbound.Host) {
+	host.CreateTime = uint64(time.Now().UnixNano())
+	r.reconcileHost(host)
+}
+
+// Reconciles the specified southbound host against its topology entity counterpart
+func (r *HostReconciler) reconcileHost(host *southbound.Host) {
+	// first, compose DeviceMAC/Port ID
+	hostID := topo.ID(host.MAC + "/" + strconv.FormatUint(uint64(host.Port), 10))
+
+	//composing IP address
+	ipAddr := topo.IPAddress{
+		IP:   host.IP,
+		Type: topo.IPAddress_IPV4,
+	}
+
+	// Try to get the host
+	_, err := r.topoClient.Get(r.ctx, &topo.GetRequest{ID: hostID})
+	if err != nil {
+		// If it is not there, create it and its relation
+		r.createHost(hostID, ipAddr, host)
+		return
+	}
+
+	// Otherwise, if it needs an update, update it
+	//r.updateHostIfNeeded(gr.Object, host, status)
+}
+
+// Register the given report, the agent ID and return....
+func (r *HostReconciler) registerReport(object *topo.Object, report *southbound.HostReport) []*southbound.Host {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// (Re)create the agent ID to device entity ID binding
+	r.agentDevices[report.AgentID] = object
+
+	// See if all hosts in the report can be processed, if not, register them in the pending hosts
+	// Otherwise, add them to the hosts to be processed now
+	hosts := make([]*southbound.Host, 0, len(report.Hosts))
+	for _, host := range report.Hosts {
+		if _, ok := r.agentDevices[host.MAC]; ok {
+			hosts = append(hosts, host)
+		}
+	}
+
+	return hosts
+}
+
+// Creates host topo object and its relation
+func (r *HostReconciler) createHost(hostID topo.ID, ipAddr topo.IPAddress, host *southbound.Host) {
+	hostAspect := &topo.NetworkInterface{MacDevicePort: host.MAC + "/" + strconv.FormatUint(uint64(host.Port), 10),
+		Ip: &ipAddr}
+	object, err := topo.NewEntity(hostID, topo.HostKind).WithAspects(hostAspect)
+	if err != nil {
+		log.Warnf("Unable to allocate host %s: %+v", hostID, err)
+		return
+	}
+	// ToDo - where/how can I obtain labels?? Do I need it at all?
+	//object.Labels = labels
+	if _, err = r.topoClient.Create(r.ctx, &topo.CreateRequest{Object: object}); err != nil {
+		log.Warnf("Unable to create host %s: %+v", hostID, err)
+		return
+	}
+
+	portID := topo.ID(strconv.FormatUint(uint64(host.Port), 10))
+	originates := topo.NewRelation(portID, hostID, topo.OriginatesKind)
+	if _, err = r.topoClient.Create(r.ctx, &topo.CreateRequest{Object: originates}); err != nil {
+		log.Warnf("Unable to create originates relation for host %s: %+v", hostID, err)
+		return
+	}
+	log.Infof("Created host %s", hostID)
+}
+
+// Updates host if the host aspect update time differs from the southbound host create time
```

Comment:
Let's delete this block of commented out code.

Reply:
Done